src/serializer/functions.js

```diff
@@ -264,34 +267,26 @@ export class Functions {
       effects = additionalFunctionEffects.effects;
       this.writeEffects.set(functionValue, additionalFunctionEffects);
 
-      // look for newly registered optimized functions
-      let modifiedProperties = additionalFunctionEffects.effects.modifiedProperties;
       // Conceptually this will ensure that the nested additional function is defined
       // although for later cases, we'll apply the effects of the parents only.
       this.realm.withEffectsAppliedInGlobalEnv(() => {
-        for (let [propertyBinding] of modifiedProperties) {
-          let descriptor = propertyBinding.descriptor;
-          if (descriptor && propertyBinding.object === optimizedFunctionsObject) {
-            let newValue = descriptor.value;
-            invariant(newValue instanceof Value); //todo: this does not seem invariantly true
-            let newEntry = this.__optimizedFunctionEntryOfValue(newValue);
-            if (newEntry) {
-              additionalFunctions.add(newEntry.value);
-              getEffectsFromAdditionalFunctionAndNestedFunctions(newEntry.value, newEntry.argModelString);
-              // Now we have to rember the stack of effects that need to be applied to deal with
-              // this additional function.
-            }
-          }
+        let newOptFuncs = this.__generateInitialOptimizedFunctionsFromRealm();
+        for (let newEntry of newOptFuncs) {
+          additionalFunctions.add(newEntry.value);
+          getEffectsFromAdditionalFunctionAndNestedFunctions(newEntry.value, newEntry.argModel);
```

Comment:
What happens when the same function is being requested to be optimized multiple times and from different nesting levels?

Reply:
What do we want to happen here?

I assume we either want to assume or verify that calling the function from all the nesting levels are the same.

Alternatively, we could just throw an error.

Reply:
Let's keep it simple: If we don't know that it will work, issue an error. Be conservative and correct.